### PR TITLE
API endpoint for bulk message actions

### PR DESCRIPTION
**Pull request by rowanseymour**
Yo review @ericnewcomer @nicpottier

### CODE REVIEW

File: temba/api/serializers.py [hunk 1]
```diff
@@ -117,6 +139,74 @@ class Meta:
...
+        if attrs[source] not in ('label', 'unlabel', 'archive', 'unarchive', 'delete'):
+            raise ValidationError("Invalid action name: %s" % attrs[source])
+        return attrs
+
+    def validate_label(self, attrs, source):
+        label_name = attrs.get(source, None)
+        if label_name:
+            label = Label.objects.filter(org=self.user.get_org(), name=label_name).first()
+            if not label:
+                label = Label.create_unique(self.user.get_org(), self.user, label_name, None)
```
Comment (rowanseymour) [inline temba/api/serializers.py]: Using `label` is different than `label_uuid` in that we'll create the label if it doesn't exist. I'd assume we'd do the same for adding contacts to a group by the group name.

Comment (nicpottier) [inline temba/api/serializers.py]: Should we add a get_or_create to Label since that seems to be our pattern?

Comment (ericnewcomer) [inline temba/api/serializers.py]: Also lookup shouldn't be case sensitive

File: temba/msgs/models.py [hunk 2]
```diff
@@ -1384,6 +1384,14 @@ class Label(TembaModel, SmartModel):
     org = models.ForeignKey(Org)
 
     @classmethod
+    def get_or_create(cls, org, user, name, parent=None):
+        label = Label.objects.filter(org=org, name__iexact=name).first()
```
Comment (nicpottier) [inline temba/msgs/models.py]: strip() too?

File: temba/msgs/views.py [hunk 3]
```diff
@@ -812,25 +812,28 @@ def get_queryset(self, **kwargs):
 
 
 class LabelForm(forms.ModelForm):
-    parent = forms.ModelChoiceField(Label.objects.all(), required=False, label=_("Parent"))
+    parent = forms.ModelChoiceField(Label.objects.none(), required=False, label=_("Parent"))
```
Comment (nicpottier) [inline temba/msgs/views.py]: OOhh, didn't know about ```none()```

File: temba/msgs/views.py [hunk 4]
```diff
@@ -879,30 +863,44 @@ class Create(ModalMixin, OrgPermsMixin, SmartCreateView):
...
+            kwargs['label'] = self.get_object()
+            return kwargs
+
+        def derive_fields(self):
+            return ('name', 'parent')
+
+    class Delete(OrgObjPermsMixin, SmartDeleteView):
+        redirect_url = "@msgs.msg_inbox"
+        cancel_url = "@msgs.msg_inbox"
+        success_message = ''
```
Comment (nicpottier) [inline temba/msgs/views.py]: Sorted by CRUD?

Comment (rowanseymour) [inline temba/msgs/views.py]: I have a habit of sorting things by their lifecycle phase... first you create, then update, finally delete.

Comment (nicpottier) [inline temba/msgs/views.py]: And on the 7th day?

Comment (rowanseymour) [inline temba/msgs/views.py]: God RESTed ?

File: temba/msgs/models.py (hunk 2 above)
Comment (rowanseymour) [inline temba/msgs/models.py]: Sure

**Pull request merged**